Add client side caching support for MultiplexedConnection

altanozlu opened the pull request:
I've cleared more over #1089 and added changes under `cache-aio` feature.
I believe it's ready to review.



Benchmark results:
```
group                       with_cache                               without_cache
-----                       ----------                               -------------
0.1-100-10000/thread-1      1.00       3.1±0.17s 320.0 KElem/sec     1.06       3.2±0.14s 302.7 KElem/sec
0.1-100-10000/thread-16     1.20   888.4±64.88ms 1099.3 KElem/sec    1.00  740.5±114.53ms 1318.8 KElem/sec
0.1-100-10000/thread-32     1.08   898.6±56.34ms 1086.8 KElem/sec    1.00   834.8±18.18ms 1169.8 KElem/sec
0.1-100-10000/thread-4      1.07  844.0±103.16ms 1157.1 KElem/sec    1.00   786.9±26.11ms 1241.0 KElem/sec
0.5-1-100000/thread-1       1.00   334.1±20.54ms 292.3 KElem/sec     1.02   339.4±30.71ms 287.7 KElem/sec
0.5-1-100000/thread-16      1.00    104.1±6.89ms 938.1 KElem/sec     1.01    104.8±7.10ms 931.7 KElem/sec
0.5-1-100000/thread-32      1.04    112.4±4.39ms 868.7 KElem/sec     1.00    108.1±3.03ms 903.4 KElem/sec
0.5-1-100000/thread-4       1.00    111.6±5.54ms 875.0 KElem/sec     1.06   118.4±10.85ms 825.1 KElem/sec
0.5-100-10000/thread-1      1.00       2.0±0.10s 487.7 KElem/sec     1.62       3.2±0.22s 302.0 KElem/sec
0.5-100-10000/thread-16     1.00   796.1±94.15ms 1226.7 KElem/sec    1.01   802.1±84.05ms 1217.5 KElem/sec
0.5-100-10000/thread-32     1.00   839.1±64.25ms 1163.8 KElem/sec    1.02   854.3±38.28ms 1143.1 KElem/sec
0.5-100-10000/thread-4      1.01   892.7±34.85ms 1094.0 KElem/sec    1.00   886.9±26.67ms 1101.1 KElem/sec
0.5-5-100000/thread-1       1.00  1126.0±118.10ms 433.7 KElem/sec    1.46  1643.4±164.08ms 297.1 KElem/sec
0.5-5-100000/thread-16      1.16   571.1±17.98ms 854.9 KElem/sec     1.00   491.3±16.86ms 993.8 KElem/sec
0.5-5-100000/thread-32      1.14   558.6±34.31ms 874.1 KElem/sec     1.00   489.6±11.13ms 997.4 KElem/sec
0.5-5-100000/thread-4       1.23   594.2±33.80ms 821.8 KElem/sec     1.00   482.0±15.11ms 1013.0 KElem/sec
0.9-100-10000/thread-1      1.00   692.4±73.42ms 1410.4 KElem/sec    4.86       3.4±0.14s 289.9 KElem/sec
0.9-100-10000/thread-16     1.00   273.8±18.87ms  3.5 MElem/sec      3.04  831.5±101.61ms 1174.5 KElem/sec
0.9-100-10000/thread-32     1.00    279.5±6.85ms  3.4 MElem/sec      3.13   873.9±38.05ms 1117.5 KElem/sec
0.9-100-10000/thread-4      1.00    230.0±4.80ms  4.1 MElem/sec      4.08   938.2±51.10ms 1040.8 KElem/sec
0.99-100-10000/thread-1     1.00    274.7±2.60ms  3.5 MElem/sec      11.73      3.2±0.22s 303.1 KElem/sec
0.99-100-10000/thread-16    1.00     69.4±4.17ms 13.7 MElem/sec      11.47  796.5±99.59ms 1226.1 KElem/sec
0.99-100-10000/thread-32    1.00     63.7±1.30ms 15.0 MElem/sec      13.83  881.5±41.41ms 1107.9 KElem/sec
0.99-100-10000/thread-4     1.00    149.8±6.65ms  6.4 MElem/sec      6.15   921.3±66.03ms 1060.0 KElem/sec
```

Commands used for benchmark:
```shell
RUSTFLAGS="-C target-cpu=native" ENABLE_CLIENT_SIDE_CACHE=false PROTOCOL=RESP3 cargo bench --all-features --bench bench_cache -- --save-baseline without_cache
RUSTFLAGS="-C target-cpu=native" ENABLE_CLIENT_SIDE_CACHE=true PROTOCOL=RESP3 cargo bench --all-features --bench bench_cache -- --save-baseline with_cache
critcmp without_cache with_cache
```
Testbench: 
- Ryzen 9950X(16C 32T) with linux 6.11.8-300.fc41.x86_64
- Redis server v=7.9.225 sha=0a24bfec:0 malloc=jemalloc-5.3.0 bits=64 build=3995de5f5cc4075a
- benchmark connects to redis server via localhost


Benchmark explanation:
`0.99-100-10000/thread-4` means 
- for each key 99% of requests will be GET and 1% of requests will be SET
- there will be 100 commands for each key
- there will be 10000 different keys
- tokio thread will have only 4 worker thread

So it means there will be 1 million requests and 10 thousands of them will be SET others are GET.

As far as I see when workload is write heavy it will be around 10% slower than non cache variant, but when it's mostly/equally read , caching will improve the performance. Having 13x performance gain for 99% read workload is nice :) Of course this is not free, client side caching will use more cpu.



I'm open to better benchmark methods 


fixes https://github.com/redis-rs/redis-rs/issues/331  https://github.com/redis-rs/redis-rs/issues/312

nihohit commented on redis/src/connection.rs:
```diff
@@ -1014,6 +1024,31 @@ pub(crate) fn connection_setup_pipeline(
+    match cache_config.mode {
+        crate::caching::CacheMode::None => {}
+        crate::caching::CacheMode::All => {
+            pipeline.cmd("CLIENT").arg("TRACKING").arg("ON").ignore();
```
I prefer that caching won't be silently ignored. If the user turns on caching with a client level <6, they should receive an informative error message.

altanozlu replied:
Now it will raise an error

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +630,63 @@ impl Cmd {
+            client_side_ttl: None,
+        });
+        self
+    }
```
nit: add empty line between methods. same elsewhere

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +630,63 @@ impl Cmd {
+        self
+    }
+    #[cfg(feature = "cache-aio")]
+    #[inline]
```
why inline this?

altanozlu replied:
My bad, it grew too much for this

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +630,63 @@ impl Cmd {
+                    });
+                }
+            } else {
+                return None;
```
nit: I think these 2 else clauses can be omitted.

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +630,63 @@ impl Cmd {
+                        is_mget = true;
+                        Some(x)
+                    }
+                    _ => return None,
```
why other commands can't be cached?

altanozlu replied:
Only read commands *should* be cached, I can prepare extensive list but this is kinda MVP.

altanozlu replied:
Hmm NOLOOP option was never on my plans, I will create an issue about it.
there was an early issue about it when it was mostly draft (https://github.com/redis-rs/redis-rs/issues/826)

altanozlu replied:
https://github.com/redis-rs/redis-rs/issues/1316

nihohit commented on redis/src/caching/statistics.rs:
```diff
@@ -0,0 +1,49 @@
+    pub invalidate: usize,
+    /// Count of commands sent when caching is enabled, only increased when key is cached.
+    pub sent_command_count: usize,
+}
```
add empty line between scopes

nihohit commented on redis/src/caching/statistics.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// CacheStatistics holds statistics generated by Client Side Caching.
+#[derive(Debug)]
+pub struct CacheStatistics {
```
Maybe `FinalizedStatistics`? I want a name that clarifies this was computed from `Statistics` and is its final form.

altanozlu replied:
I chose `CacheStatistics` because `Statistics` isn't a public struct and users don't have to know about it. So `FinalizedStatistics` might bring more confusion.

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// Holds configuration to be used by [`crate::caching::CacheManager`]
+#[derive(Clone, Debug, Copy)]
+pub struct CacheConfig {
```
please use the builder pattern instead of exposing internal fields. That way we can refactor the class without breaking user code.

altanozlu replied:
I've stopped exposing internal field and added setter methods.

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,49 @@
+#[derive(Clone, PartialEq, Debug, Copy)]
+pub enum CacheMode {
+    /// Don't cache
+    None,
```
why is there a `None` mode? wouldn't a disabled cache just be a `let cache_config: Option<CacheMode> = None;`?

altanozlu replied:
Hmm I will refactor CacheConfig part in general, that might be better

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,49 @@
+
+impl CacheConfig {
+    /// Enables cache with 10_000 LRU size and 30 minutes of client ttl
+    pub fn enabled() -> Self {
```
this looks like a `Default` implementation. the name doesn't show that this is just a proposed default.

altanozlu replied:
With changes to cacheconfig default is now enabled.

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -88,14 +96,26 @@ impl Debug for Pipeline {
-        error: Option<RedisError>,
-        push_sender: Option<AsyncPushSender>,
-    }
+struct PipelineSink<T> {
```
please fix indentation.

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -88,14 +96,26 @@ impl Debug for Pipeline {
+    push_sender: Option<AsyncPushSender>,
+    cache_manager: CacheManager,
+}
+}
```
add empty line

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -164,8 +190,19 @@ where
             Ok(Value::Push { kind, data }) if !kind.has_reply() => {
+                #[cfg(feature = "cache-aio")]
+                if kind == PushKind::Invalidate {
+                    if let Some(Value::Array(redis_key)) = data.first() {
```
I assume that this logic will be repeated in other connection types. It probably shouldn't exist here.

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -164,8 +190,19 @@ where
+                    }
+                }
                 send_push(self_.push_sender, PushInfo { kind, data });
-
```
why did you remove all of these empty lines?

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -520,15 +580,180 @@ impl MultiplexedConnection {
+        cmd: &Cmd,
+        ci: &CommandCacheInformationByRef<'a>,
+    ) -> RedisResult<Value> {
+        let mut response: Vec<Value> = vec![];
```
please document what this function does. Its huge and unclear.

altanozlu replied:
I've documented and refactored it, I believe it's better now, can you check it again please ?

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +630,63 @@ impl Cmd {
+                        client_side_ttl: self.cache.as_ref().and_then(|x| x.client_side_ttl),
+                        redis_key: key,
+                        cmd: if is_mget { x } else { self.data.as_slice() },
+                        is_mget,
```
mget isn't the only multi-key command. for example, there's mset.

altanozlu replied:
but MSET is not something can be cached, it will trigger invalidation. There can be also JSON.MGET but I'm not sure if there are other ones for client side caching.

altanozlu replied:
Also Added JSON.MGET to here but didn't implement it yet.

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -520,15 +580,180 @@ impl MultiplexedConnection {
+            CacheMode::None => None,
+        };
+        #[cfg(feature = "cache-aio")]
+        let (cmd_bytes, cmd_count, notifier) = if let Some(ci) = &cache_information {
```
there's a lot of code here and above which isn't multiplexer specific, and seems to also be relevant for other connections - anything which involves creating new commands, for example. Please move that to the cache module.

altanozlu replied:
I've moved some but not all. Is this enough ? I prefer not moving everything to cache module when potential overlap is unknown.

nihohit reviewed (commented):
Great start!
I didn't go over all of the code yet, so these are just the initial comments.

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,226 @@
+    const MAX_SHARD_SIZE: usize = 32;
+    fn new(total_key_size: NonZeroUsize) -> Self {
+        // If total cache size is smaller than max shard size then it won't use sharding.
+        let (shard_size, per_lru_size) = if total_key_size.get() >= Self::MAX_SHARD_SIZE {
```
is `shard_size` actually `shard_count`? otherwise I don't understand what's the difference between `shard_size` and `per_lru_size`.

altanozlu replied:
I've changed it and now it should be clear.

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,226 @@
+struct CacheCmdEntry {
+    cmd: RedisCmd,
+    value: Option<Value>,
+    receiver: Receiver<Value>,
```
what is this? what purpose does it serve?

altanozlu replied:
If multiple request is trying to access same key with same command, only one request will send request to server, other requests will wait this  `receiver`
```
if receiver.changed().await.is_ok() {
                self.statistics.increase_hit(1);
                return Ok(Some(receiver.borrow().clone()));
           }
```

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,226 @@
+}
+
+struct CacheItem {
+    ttl: Instant,
```
nit: time to live is a duration. if this is an instant, it's `expiry_time`, or some similar name.

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,226 @@
+        cache_information: &CommandCacheInformationByRef<'a>,
+    ) -> Result<Option<Value>, Sender<Value>> {
+        let mut receiver = {
+            let mut lru_cache_better = self
```
what do you mean by better?

altanozlu replied:
oops, will be removed (it's coming from an old patch)

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,226 @@
+                .lock()
+                .unwrap();
+            if let Some(ch) = lru_cache_better.get_mut(cache_information.redis_key) {
+                // We have found redis key in cache, but we need to make sure redis cmd is also here.
```
`is also here` - what do you mean by "here"?

altanozlu replied:
I've updated the documentation.

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,226 @@
+                .get_shard(cache_information.redis_key)
+                .lock()
+                .unwrap();
+            if let Some(ch) = lru_cache_better.get_mut(cache_information.redis_key) {
```
this function is fairly large, mostly because of large conditional split. Can it be torn apart to smaller, clearly named functions?

altanozlu replied:
I will work on this

altanozlu replied:
I've split the function, it required new enums for handling the logic

altanozlu commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,268 @@
+            .await
+            .unwrap();
+        assert_eq!(val, None);
+        {
```
It's a precaution I use when there is multiple rebinding to the same name, I can remove them.

nihohit commented on redis/tests/support/mod.rs:
```diff
@@ -549,6 +550,39 @@ impl TestContext {
     }
 
+    #[cfg(all(feature = "tokio-comp", feature = "cache-aio"))]
+    pub fn multiplexed_async_connection_tokio_with_cache(
```
why is this needed? I see that everywhere you use `multiplexed_async_connection_tokio_with_cache_config` it's with the default cache.

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+mod support;
+
+#[test]
+fn test_cache() {
```
I'm thinking about implementing a mock multiplexed connection, would that make the statistics redundant?

altanozlu replied:
I want to test with real connection, because how client side caching works is really about how redis server responds. This way we might be able to see any issue comes up new valkey,redis versions.

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+        }
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        assert_eq!(s.miss, 1000);
```
why is this?
shouldn't some of the concurrent operations hit the cache, like in line 41?

altanozlu replied:
What is being tested here is 1000 future is trying to get same key, and one of them will be able to send the query and others will hold guard until that query is completed. Since the one tries to send query will timeout others will stop holding guard and query the server instead. 1 for first miss from first future sends the query 999 miss from futures awaited the guard and dropped it.

I guess this one requires more wording on what it does.

nihohit replied:
yes, sorry - that wasn't at all clear from the test :)

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+            let mut c = con.clone();
+            v.push(async move {
+                let x: RedisFuture<Option<u32>> = c.get("KEY");
+                tokio::time::timeout(Duration::from_micros(i), x).await
```
why is the timeout needed? just for safety?

altanozlu replied:
Answered below.

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+        let mut v = vec![];
+        for _ in 0..5 {
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
```
why are the spawns needed?

altanozlu replied:
For more reliability in tests, to make sure it doesn't rely in order of commands. Can be removed but I see no harm.

nihohit replied:
this is fine - just add a comment explaining this, please.

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+}
+
+#[test]
+fn test_cache_guards_and_different_types() {
```
test_cache_is_not_target_type_dependent?

altanozlu replied:
CacheManager stores `Value` and a `Value::SimpleString` can be converted into `String` or `i32`. This pattern also aligns with normal command usage.

nihohit replied:
right, I suggest what I think is a clearer test name. sorry I wasnt' clear :)

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        // It should get value once and miss then it should use cache.
+        assert_eq!(s.hit, 14);
```
why here the requests hit the cache and in `test_cache_drops_and_guards` they don't?

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,307 @@
+}
+
+#[test]
+fn test_cache_mget_verify() {
```
I'm having a hard time understanding the purpose of all of the tests. The names are very terse, and require knowledge of the internal implementation in order to understand what they're actually testing.

altanozlu replied:
I will add more documentation and make them clearer.

nihohit reviewed (commented):
Sorry that this is taking forever. I'm trying to review this in chunks when I have the time, but the last couple of weeks have been intensive, and the chunks are large :(

nihohit commented:
I briefly scanned through the multiplexed connection changes, and noticed these broad issues:
1. caching is only used when sending single commands, not pipelines.
2. there are very few usage of `self` that aren't `self.cache_manager` in the logic surrounding caching. That implies that most of the logic there could be moved to the cache manager, and tested without relying on the TCP connection.
3. Couldn't the interface between the connection and the cache manager be something like 
* pass command/pipeline to cache manager.
* cache manager returns an enum with one of these variants:
- `Value` for cmd / `Vec<Value>` for pipeline, if all values are cached
- `Cmd`/`Pipeline`, in case no value was cached. This can either be the passed command, or a new pipeline generated to enable caching.
- `(Vec<Value>, Cmd, Instructions)` / `<Vec<Value>, pipeline, Instructions)` in case some of the values are cached and some aren't. `Instructions` is an object explaining how to interleave the returned values.
* connection sends the received command, and updates the cache with its result.

This means that the cache manager needs to only expose 2 functions to the connection (not including the functions exposed to the pipeline) 
1. get_cached_value_or_command
2. set_command_result 
WDYT? am I missing something?

altanozlu commented:
> I briefly scanned through the multiplexed connection changes, and noticed these broad issues:
> 
>     1. caching is only used when sending single commands, not pipelines.
> 
>     2. there are very few usage of `self` that aren't `self.cache_manager` in the logic surrounding caching. That implies that most of the logic there could be moved to the cache manager, and tested without relying on the TCP connection.
> 
>     3. Couldn't the interface between the connection and the cache manager be something like
> 
> 
>     * pass command/pipeline to cache manager.
> 
>     * cache manager returns an enum with one of these variants:
> 
> 
>     * `Value` for cmd / `Vec<Value>` for pipeline, if all values are cached
> 
>     * `Cmd`/`Pipeline`, in case no value was cached. This can either be the passed command, or a new pipeline generated to enable caching.
> 
>     * `(Vec<Value>, Cmd, Instructions)` / `<Vec<Value>, pipeline, Instructions)` in case some of the values are cached and some aren't. `Instructions` is an object explaining how to interleave the returned values.
> 
> 
>     * connection sends the received command, and updates the cache with its result.
> 
> 
> This means that the cache manager needs to only expose 2 functions to the connection (not including the functions exposed to the pipeline)
> 
>     1. get_cached_value_or_command
> 
>     2. set_command_result
>        WDYT? am I missing something?

I had few try for managing everything (at least for MGET) in cache manager and just using mpx for executing commands, I've eventually scrapped them because return part became a complex enum and cachemanager became state manager. It felt like an overkill. If you are also okay with complex enum passed between mpx and cachemanager, I would like to work on it again.

About first point (single commands) I didn't want to deal with pipelines for first iteration, if it's you see something as a *must* have, I'd like to work on it, it may just take more time...

altanozlu commented on redis/src/caching/statistics.rs:
```diff
@@ -0,0 +1,53 @@
+
+/// CacheStatistics holds statistics generated by Client Side Caching.
+#[derive(Debug, Default)]
+pub struct CacheStatistics {
```
For sure it makes easier to test things, but my main goal for this to be used by users. As an user if I have enabled client side caching in my application, I have no idea how well it works rather than seeing latency decrease etc. But by using CacheStatistics I can generate metrics from it.

nihohit commented:
> About first point (single commands) I didn't want to deal with pipelines for first iteration

That's a good question. If we didn't expose the statistics, then I'd say that as far as the users are concerned, it doesn't matter. But because we do expose statistics, I think that it would be confusing for the users if some requests changed statistics and some didn't. 
I definitely wouldn't want to ship a version without pipeline support, but maybe we could merge the initial PR and then either finish with pipeline support before releasing a version, or hide the caching API, release a version, and continue on pipelines.

>  If you are also okay with complex enum passed between mpx and cachemanager, I would like to work on it again.

generally I'm ok with complex enums, but do you have a measure of how complex the enum will need to be?

altanozlu commented:
> > About first point (single commands) I didn't want to deal with pipelines for first iteration
> 
> That's a good question. If we didn't expose the statistics, then I'd say that as far as the users are concerned, it doesn't matter. But because we do expose statistics, I think that it would be confusing for the users if some requests changed statistics and some didn't. I definitely wouldn't want to ship a version without pipeline support, but maybe we could merge the initial PR and then either finish with pipeline support before releasing a version, or hide the caching API, release a version, and continue on pipelines.
> 
> > If you are also okay with complex enum passed between mpx and cachemanager, I would like to work on it again.
> 
> generally I'm ok with complex enums, but do you have a measure of how complex the enum will need to be?

If pipeline support is important for the release then I can batch them up. I won't be able to work on this for next few weeks, I will publish revised patch in november.

nihohit commented:
Thanks, you're doing awesome work!
If you think that pipeline support is more of the same, than we can finish with the current design and then implement the pipeline on top of it. I think that you were right to limit the initial work to single commands, but as a feature, it isn't complete without pipelines IMO.

altanozlu commented:
Hey! I've rebased over main and changed few things:

- Splitted LRU related code from CacheManager into sharded_lru.rs
- Got rid of guard logic in order to reduce complexity, I want to add back again but not in this PR.
- Created CacheablePipeline and CacheableCommand to store states for requests.
- Unified cache and cmd creation logic for cmd/pipeline under caching/cache_manager.rs , caching/cmd.rs instead of using mpx directly.
- Updated tests for rstest


This is not final commit, I'm planning to add more tests and Optout support.
Until then 👋🏻

altanozlu commented:
Updated benchmark results, I'm planning to add more benchmarks in upcoming PRs, @nihohit can I get a review ?

nihohit commented on redis/src/aio/multiplexed_connection.rs:
```diff
@@ -556,6 +646,16 @@ impl MultiplexedConnection {
+
+    /// Gets [`crate::caching::CacheStatistics`] for current connection.
+    #[cfg(feature = "cache-aio")]
+    pub fn get_cache_statistics(&self) -> CacheStatistics {
```
nit: this should probably be an `Option<CacheStatistics>`, so that `None` is returned if there's no cache manager. (and adjust docs)

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +615,38 @@ impl Cmd {
+    /// It marks command as triable for client side caching.
+    /// Useful when `CacheMode` is set to `CacheMode::OptIn`
+    #[cfg(feature = "cache-aio")]
+    pub fn cache(&mut self) -> &mut Cmd {
```
IMO this should take `client_side_ttl` and remove `cache_with_client_side_ttl`.
Do you think that future iterations might take more arguments? should we create some config struct for per-command caching, in order to maintain forward compatibility.

altanozlu replied:
Creating option struct and having one function to take it makes more sense because currently this function also enables for OptIn mode and once OptOut is introduced there will be need for another function.

nihohit commented on redis/src/pipeline.rs:
```diff
@@ -350,3 +352,24 @@ macro_rules! implement_pipeline_commands {
+    /// It marks command as triable for client side caching with custom Client Side TTL for cache.
+    /// Useful when `CacheMode` is set to `CacheMode::OptIn`
+    #[cfg(feature = "cache-aio")]
+    pub fn cache_with_client_side_ttl(&mut self, client_side_ttl: Duration) -> &mut Self {
```
same as for `Cmd` re: not having 2 functions.

nihohit commented on redis/src/caching/cache_manager.rs:
```diff
@@ -0,0 +1,215 @@
+}
+
+impl CacheManager {
+    pub fn new(cache_config: CacheConfig) -> Self {
```
nit: the `pub` functions here should be `pub(crate)`

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,60 @@
+use std::num::NonZeroUsize;
+use std::time::Duration;
+
+/// Defines CacheMode of CacheConfig
```
Defines the behavior of the cache regarding which commands should be cached.

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,60 @@
+/// Defines CacheMode of CacheConfig
+#[derive(Clone, PartialEq, Debug, Copy)]
+pub enum CacheMode {
+    /// Try to cache every command.
```
Try to cache every command. This saves the user a call of [Cmd::cache].

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,60 @@
+    /// Try to cache every command.
+    All,
+    /// Only try to cache commands which explicitly called cache methods.
+    /// If OptIn is used `cache` or `cache_with_client_side_ttl` functions
```
use [Cmd::cache] to link to the function.

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,60 @@
+    OptIn,
+}
+
+/// Options for managing client side caching
```
Configuration for client side caching.

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,60 @@
+        self
+    }
+
+    ///  Sets the default client side TTL when it's not set with `cache_with_client_side_ttl` per command.
```
try to avoid acronyms in documentation, or at least add the expanded version - 
Sets the default client-side Time To Live (TTL) for cached values, when a TTL isn't explicitly passed on a `Cmd`.

nihohit reviewed (commented):
public API is good. I'll try to get another round on the `caching` module. One thing that I already noticed is that there are no unit tests there.

nihohit commented:
BTW, the benchmark is for a local server, right? I assume that the number will be better for the cache once you have to pay for network wait, too.

altanozlu commented:
> BTW, the benchmark is for a local server, right? I assume that the number will be better for the cache once you have to pay for network wait, too.

Indeed I'm planning to use `tc` for simulating 1ms latency environment.
Will look into rest of the review tomorrow

altanozlu commented:
Test coverage was already 88% with existing tests, now I've also added unit tests for areas tests didn't cover.
Also applied changes in the review

altanozlu commented:
I will dig into more about PipelineResponseExpectation for now just pushed to pass CI

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,438 @@
+            assert_hit(&con, 0);
+            assert_miss(&con, 2);
+
+            let _: Vec<Option<String>> = redis::cmd("MGET")
```
can this check for keys 1,2, &3, but in the order of 1,3,2?
Just to verify that the caching mechanism is not in some way order dependent

altanozlu replied:
Changed the order also added SET and asserts for values.

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,438 @@
+            assert_hit(&con, 0);
+            assert_miss(&con, 2);
+
+            let ((k1, k2), k3, k4): ((i32, i32), i32, Option<i32>) = get_pipe(atomic)
```
nit: k1-4 implies *key* 1-4, but you're only testing 2 keys. maybe res1-4.

altanozlu replied:
Changed them to r* also there is something similar in test_cache_pipeline_partial_opt_in, I will keep it unresolved for now.

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,438 @@
+                .arg(&["key_1", "key_2"])
+                .cmd("GET")
+                .arg("key_1")
+                .cmd("GET")
```
can this be moved one command earlier, just to verify that the mechanism isn't order-dependent?

altanozlu replied:
I've put MGET in between them

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,438 @@
+#[rstest]
+#[case::tokio(RuntimeType::Tokio)]
+#[cfg_attr(feature = "async-std-comp", case::async_std(RuntimeType::AsyncStd))]
+fn test_cache_opt_in(#[case] runtime: RuntimeType) {
```
it's unclear from the name of the test how this is different from `test_cache_basic` with opt_in=true

altanozlu replied:
Parts with `// Since cache is not used, hit should still be 0` and `// GET is not marked with cache(), there should be no MISS/HIT`  which is important to check OptIn works as opt in

altanozlu replied:
Tried my best to rename it but I'm bad at this :/

nihohit commented on redis/tests/test_cache.rs:
```diff
@@ -0,0 +1,438 @@
+#[rstest]
+#[case::tokio(RuntimeType::Tokio)]
+#[cfg_attr(feature = "async-std-comp", case::async_std(RuntimeType::AsyncStd))]
+fn test_cache_pipeline_with_opt_in(
```
same

nihohit reviewed (commented):
love the tests!

nihohit commented:
a strategic question - I don't think that we should ship this as part of 0.28 - that version will concentrate on pubsub, and I don't want to burden it with another large feature.
Will we wait with merging this until after the version ships? merge it but somehow disable/hide the feautre?
I don't want to use another long living feature branch - that was pretty painful with RESP3 support. WDYT?

altanozlu commented:
using cfg and putting it behind unstable flag like tokio does as temporary measure could work but doesn't look nice. So it's better to merge it after releasing 0.28. And maybe cut out ` add pubsub support to async cluster - TBD` from 0.28 plans to make things faster ? Let me know if you know better way to hide a feature.

nihohit commented:
no, that's the only way I know - using configuration flags.

wojciechbator commented:
You could still push it behind the flag in 0.28 and add a note of experimental support for client side caching. Then, in next release you could add it as a part of the main lib, and that'd still be a major introduction for most of users. A benefit of this would be a feedback from early users and less headache later.

altanozlu commented on redis/Cargo.toml:
```diff
@@ -121,6 +123,8 @@ sentinel = ["dep:rand"]
 tcp_nodelay = []
 num-bigint = []
 disable-client-setinfo = []
+hashbrown = ["dep:hashbrown"]
```
It looks like it's removed in https://github.com/redis-rs/redis-rs/commit/a871419f53381dab429e596c6dc50b2d09bd711c#diff-0024a66d093c343a9c17108e393c8ebb89213d963deaeba6a7547ab7c799304dL126

but somehow it got here via rebase

altanozlu commented:
@nihohit I'll push another patch soon regarding hashbrown and adding experimental clause in docs. Then it's good to go.

nihohit commented on redis/src/lib.rs:
```diff
@@ -101,6 +101,7 @@
 //! * `uuid`: enables type conversion to UUID (optional)
 //! * `sentinel`: enables high-level interfaces for communication with Redis sentinels (optional)
 //! * `json`: enables high-level interfaces for communication with the JSON module (optional)
+//! * `cache-aio`: enables **experimental** client side caching for MultiplexedConnection and ConnectionManager (optional)
```
still not for connection manager :(

altanozlu replied:
ah :(

nihohit commented on redis/src/cmd.rs:
```diff
@@ -601,6 +659,20 @@ impl Cmd {
+
+    #[cfg(feature = "cache-aio")]
+    #[inline]
+    pub(crate) fn get_cache(&self) -> &Option<CommandCacheConfig> {
```
get_cache_config

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,61 @@
+#[derive(Clone, PartialEq, Debug, Copy)]
+pub enum CacheMode {
+    /// Try to cache every command. This is equivalent to calling [crate::Cmd::set_cache_config]/[crate::Pipeline::set_cache_config] on every command.
+    /// Disabling individual commands for caching with this [CacheMode::All] will not work.
```
"Disabling individual commands for caching" - what does this mean? Do you mean "Disabling caching for individual commands"?

altanozlu replied:
Changed also here

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,61 @@
+/// Defines the behavior of the cache regarding which commands should be cached.
+#[derive(Clone, PartialEq, Debug, Copy)]
+pub enum CacheMode {
+    /// Try to cache every command. This is equivalent to calling [crate::Cmd::set_cache_config]/[crate::Pipeline::set_cache_config] on every command.
```
Try to cache the key & value for every relevant command? I think that I'm missing a document that explains what does caching actually mean - it's not commands that are cached, it's keys from sent commands, right?
I can write the explanation, if you prefer.

altanozlu replied:
It's actually both key and command, when you try `GET mouse` it will create new CacheItem and put it on LRU using key `mouse` then it will write `GET A` information and it's response under CacheItem's value_list (Vec). When you use another command like `STRLEN mouse` it will not use cached one because `STRLEN mouse` is not cached `GET mouse` is. Holding all command instead of  cmd name makes sense for cases like `GETRANGE mouse 0 3`.

altanozlu replied:
I will write more specific information in documentation.

nihohit replied:
thank you :)

nihohit replied:
what's the rationale behind this? I would like to cache `GET mouse`, and get the cached value from `MGET mouse` too.

altanozlu replied:
for MGET/ JSON.MGET it works like you expected, they have special case and they will be saved&checked as their singular form GET/JSON.GET.

you can check `test_cache_mget` for this and I can add more unit tests in future PRs regarding this.

nihohit replied:
But what's the reason for this? Why cache the value for mget but not for
strlen?

altanozlu replied:
If `STRLEN A` is sent then result of it will be cached but `GET A` won't use that cached information, because their command is different and it make sense because otherwise you can get response of `STRLEN A` when you send `GET A` which is broken. MGET is just a safe optimization.
If you think this still doesn't make sense can you please explain what  should be cached for STRLEN exactly?

nihohit replied:
oh, ok, I see - it's because the response of STRLEN is the computed value in the server. Sorry, this flew over my head 🤦.

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,61 @@
+    /// Try to cache every command. This is equivalent to calling [crate::Cmd::set_cache_config]/[crate::Pipeline::set_cache_config] on every command.
+    /// Disabling individual commands for caching with this [CacheMode::All] will not work.
+    All,
+    /// Only try to cache commands which explicitly enabled caching.
```
Only cache commands* that are explicitly marked by the user for caching?

or "keys from commands"

nihohit commented on redis/src/caching/config.rs:
```diff
@@ -0,0 +1,61 @@
+        self
+    }
+
+    ///  Sets the default client side time to live (TTL) for cached values, when a TTL isn't explicitly passed to [crate::Cmd]/[crate::Pipeline].
```
The doc here should explain how to explicitly pass a TTL to a command. Also, since `CacheMode::All` is like setting a cache config on every command, which TTL take precedence when using that cache mode?

altanozlu replied:
> The doc here should explain how to explicitly pass a TTL to a command.
will do it with next patch

> Also, since CacheMode::All is like setting a cache config on every command, which TTL take precedence when using that cache mode?

`   ///  Sets the default client side time to live (TTL) for cached values, when a TTL isn't explicitly passed to [crate::Cmd]/[crate::Pipeline].`
It actually says that if user isn't setting it explicitly, default client side ttl will be used. I will reword this bit

nihohit commented on redis/src/caching/mod.rs:
```diff
@@ -0,0 +1,18 @@
+//! This module provides **experimental** support for Server-assisted client-side caching in Redis.
+//!
+//! This implementation of Server-assisted client-side cache, only works with RESP3 connections.
+//! Some Redis providers might not enable this feature (e.g. Google Cloud Memorystore).
```
lt's not name `Memorystore` here, since this will require us to update the docs if they change their behavior.

nihohit commented on redis/src/caching/mod.rs:
```diff
@@ -0,0 +1,18 @@
+//! This implementation of Server-assisted client-side cache, only works with RESP3 connections.
+//! Some Redis providers might not enable this feature (e.g. Google Cloud Memorystore).
+//! Commands that are cached will be kept in memory, commands won't be deleted from memory immediately,
+//! instead they will be removed by LRU or when same key is requested again.
```
by LRU, by their Time To Live (TTL), or when ...

altanozlu replied:
I'm changing the wording and also adding redis invalidation.

nihohit reviewed (approved):
still didn't have time to look at the internal implementation, but aside from the doc corrections, I think this is ready to merge, and we'll iterate on the internal implementation from main.
Very through work, well done!

nihohit commented:
But what's the reason for this? Why cache the value for mget but not for
strlen?

On Tue, 28 Jan 2025, 22:37 Altan Ozlu, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In redis/src/caching/config.rs
> <https://github.com/redis-rs/redis-rs/pull/1296#discussion_r1932826405>:
>
> > @@ -0,0 +1,61 @@
> +use std::num::NonZeroUsize;
> +use std::time::Duration;
> +
> +/// Defines the behavior of the cache regarding which commands should be cached.
> +#[derive(Clone, PartialEq, Debug, Copy)]
> +pub enum CacheMode {
> +    /// Try to cache every command. This is equivalent to calling [crate::Cmd::set_cache_config]/[crate::Pipeline::set_cache_config] on every command.
>
> for MGET/ JSON.MGET it works like you expected, they have special case and
> they will be saved&checked as their singular form GET/JSON.GET.
>
> you can check test_cache_mget for this and I can add more unit tests in
> future PRs regarding this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redis-rs/redis-rs/pull/1296#discussion_r1932826405>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDEBBVPSTK4A7D3QQ77G2L2M7S7LAVCNFSM6AAAAABNN7FQAGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZZGM3DGOJRGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

nihohit commented:
Excellent work @altanozlu. Anything to add before we merge this?

altanozlu commented:
> Excellent work @altanozlu. Anything to add before we merge this?

Currently none, there are some changes regarding benchmark but I would like to do it in next iterations.

nihohit commented:
Then it's in! Hurrah!
